internal/cloud/realm/user.go

```diff
@@ -92,11 +93,139 @@ type User struct {
 // UserIdentity is a Realm application user identity
 type UserIdentity struct {
 	UID          string                 `json:"id"`
-	ProviderType string                 `json:"provider_type"`
+	ProviderType ProviderType           `json:"provider_type"`
 	ProviderID   primitive.ObjectID     `json:"provider_id"`
 	ProviderData map[string]interface{} `json:"provider_data,omitempty"`
 }
 
+// ProviderType is a Realm application provider type
+type ProviderType string
+
+// set of supported provider type values
+const (
+	ProviderTypeEmpty          ProviderType = ""
+	ProviderTypeUserPassord    ProviderType = "local-userpass"
+	ProviderTypeAPIKey         ProviderType = "api-key"
+	ProviderTypeFacebook       ProviderType = "oauth2-facebook"
+	ProviderTypeGoogle         ProviderType = "oauth2-google"
+	ProviderTypeAnonymous      ProviderType = "anon-user"
+	ProviderTypeCustomToken    ProviderType = "custom-token"
+	ProviderTypeApple          ProviderType = "oauth2-apple"
+	ProviderTypeCustomFunction ProviderType = "custom-function"
+)
+
+// slice of supported provider type values as string slice
+var (
+	ValidProviderTypes = []string{
+		ProviderTypeUserPassord.String(),
+		ProviderTypeAPIKey.String(),
+		ProviderTypeFacebook.String(),
+		ProviderTypeGoogle.String(),
+		ProviderTypeAnonymous.String(),
+		ProviderTypeCustomToken.String(),
+		ProviderTypeApple.String(),
+		ProviderTypeCustomFunction.String(),
+	}
+)
+
+// String returns the provider type string
+func (pt ProviderType) String() string { return string(pt) }
+
+// Display returns the provider type display string
+func (pt ProviderType) Display() string {
+	switch pt {
+	case ProviderTypeAnonymous:
+		return "Anonymous"
+	case ProviderTypeUserPassord:
+		return "User/Password"
+	case ProviderTypeAPIKey:
+		return "ApiKey"
+	case ProviderTypeApple:
+		return "Apple"
+	case ProviderTypeGoogle:
+		return "Google"
+	case ProviderTypeFacebook:
+		return "Facebook"
+	case ProviderTypeCustomToken:
+		return "Custom JWT"
+	case ProviderTypeCustomFunction:
+		return "Custom Function"
+	default:
+		return "Unknown"
+	}
+}
+
+// IsValid validates the provided provider type
+func (pt ProviderType) IsValid() error {
```

Comment:
[nit] the function name here makes me think the function should return `bool` here.  is the error actually needed/useful here (i can't find any usages here)

Reply:
I prefer to use error in situations where the type has validation on it. If it was more optional I tend to agree on going with bool.

Reply:
Point being if you feel the error is necessary, let's rename the function to something like `Validate`.  If you want to keep it as `IsValid`, then we should stick to returning a `bool` here.

But my bigger question is: what uses this / why is this needed?  I cannot find any usages of it outside of a test file.  Let's remove it all together if we don't need it.  Otherwise, it'd be nice to keep it similar to how we've approached this pattern with `UserState` above (and keep the `isValidUserState` function unexported and used only here)

Reply:
talked offline and decided to remove